daemon/oci_windows.go

```diff
@@ -183,9 +193,72 @@ func (daemon *Daemon) createSpec(c *container.Container) (*libcontainerd.Spec, e
 			Iops: &c.HostConfig.IOMaximumIOps,
 		},
 	}
+
+	// Add specific Windows credentials
+	if len(c.HostConfig.CredentialSpec) > 0 {
+		v := strings.ToLower(c.HostConfig.CredentialSpec)
+		f := strings.SplitAfter(v, "file://")
+		r := strings.SplitAfter(v, "registry://")
+		if len(f) == 2 {
+			f[1] = filepath.Clean(f[1])
+			if filepath.IsAbs(f[1]) {
+				return nil, fmt.Errorf("invalid credential spec - file:// path cannot be absolute")
+			}
+			base := filepath.Join(daemon.root, credentialSpecFileLocation)
+			full := filepath.Join(base, f[1])
+			if !strings.HasPrefix(full, base) {
+				return nil, fmt.Errorf("invalid credential spec - file:// path must be under %s", base)
+			}
+			s.Windows.CredentialSpec = readCredentialSpecFile(c.ID, full)
+		} else if len(r) == 2 {
+			s.Windows.CredentialSpec = readCredentialSpecRegistry(c.ID, r[1])
+		} else {
+			return nil, fmt.Errorf("invalid credential spec - must be prefixed file:// or registry://")
+		}
+	}
+
 	return (*libcontainerd.Spec)(&s), nil
 }
 
+// readCredentialSpecRegistry is a helper function to read a credential spec from
+// the registry. If not found, we return an empty string and warn in the log.
+// This allows for staging on machines which do not have the necessary components.
+func readCredentialSpecRegistry(id, name string) string {
+	var (
+		k   registry.Key
+		err error
+		val string
+	)
+	if k, err = registry.OpenKey(registry.LOCAL_MACHINE, credentialSpecRegistryLocation, registry.QUERY_VALUE); err != nil {
+		logrus.Warnf("%s could not be opened. Ignoring credential spec %q for container %s", credentialSpecRegistryLocation, name, id)
```

Comment:
Is it really just a warning? I wonder if we're not better of failing here that waiting for a potential (likely?) runtime failure as credentials happen to be missing.